include/rnp/rnp2.h

```diff
@@ -0,0 +1,241 @@
+* @param msg_len the length of msg in bytes
+* @param sig on success, the output signature buffer
+* @param sig_len on success, the length of sig in bytes
+*/
+rnp_result rnp_sign(rnp_private_key    key,
+                    pgp_hash_alg_t     hash,
+                    rnp_signature_type type,
+                    const uint8_t      msg[],
+                    size_t             msg_len,
+                    uint8_t **         sig,
+                    size_t *           sig_len);
+
+/**
+* Verify a signature with embedded message
+* @param key the public key
+* @param sig the signature
+* @param sig_len length of signature in bytes
+* @param msg on succes, output buffer to the message
+* @param msg_len on success, length of msg in bytes
+* @param hash_used if non-null and on success, set to the
+* hash algorithm used to create the signature.
+*/
+rnp_result rnp_verify(rnp_public_key  key,
+                      const uint8_t   sig[],
+                      size_t          sig_len,
+                      uint8_t **      msg,
+                      size_t *        msg_len,
+                      pgp_hash_alg_t *hash_used);
+
+/**
+* Verify a detached signature
+* @param key the key to verify with
+* @param msg the message bits being verified
+* @param msg_len length of msg in bytes
+* @parma sig the signature
+* @param sig_len length of signature in bytes
+* @param hash_used if non-null, and the signature is verified,
+* the hash algorithm used to create the signature will be returned.
+* @return RNP_SUCCESS if valid otherwise error
+*/
+rnp_result rnp_verify_detached(rnp_public_key  key,
+                               const uint8_t   msg[],
+                               size_t          msg_len,
+                               const uint8_t   sig[],
+                               size_t          sig_len,
+                               pgp_hash_alg_t *hash_used);
+
+/* Encryption/decryption operations */
+
+rnp_result rnp_encrypt(rnp_public_key key,
+                       pgp_symm_alg_t cipher,
+                       const uint8_t  msg[],
+                       size_t         msg_len,
+                       bool           armor,
+                       uint8_t **     output,
+                       size_t *       output_len);
+
+/**
+* Decrypt a message
+* @param key the private key to attempt decryption with
+* @param msg the ciphertext
+* @param msg_len length of msg in bytes
+* @param output pointer that will be set to a newly allocated
+* buffer, length *output_len, free with rnp_buffer_free
+* @param output_len will be set to the length of output
+* @param cipher_used if non-null will be set to the cipher used
+* to encrypt the message.
+*/
+rnp_result rnp_decrypt(rnp_private_key key,
+                       const uint8_t   msg[],
+                       size_t          msg_len,
+                       uint8_t **      output,
+                       size_t *        output_len,
+                       pgp_symm_alg_t *cipher_used);
```

Comment:
This one may be more difficult than this. The user may not know what key was used to encrypt something and instead would want the API to search some keyring for the appropriate seckey (and possibly request password for that seckey if needed).

Reply:
Agree, this could require some callbacks.

Reply:
Yes we need callbacks 👍